bundles/org.openhab.binding.evcc/src/main/java/org/openhab/binding/evcc/internal/EvccHandler.java

```diff
@@ -0,0 +1,547 @@
+import java.util.concurrent.TimeUnit;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.openhab.binding.evcc.internal.dto.Loadpoint;
+import org.openhab.binding.evcc.internal.dto.Status;
+import org.openhab.core.io.net.http.HttpUtil;
+import org.openhab.core.library.types.DateTimeType;
+import org.openhab.core.library.types.DecimalType;
+import org.openhab.core.library.types.OnOffType;
+import org.openhab.core.library.types.QuantityType;
+import org.openhab.core.library.types.StringType;
+import org.openhab.core.library.unit.MetricPrefix;
+import org.openhab.core.library.unit.SIUnits;
+import org.openhab.core.library.unit.Units;
+import org.openhab.core.thing.Channel;
+import org.openhab.core.thing.ChannelUID;
+import org.openhab.core.thing.Thing;
+import org.openhab.core.thing.ThingStatus;
+import org.openhab.core.thing.ThingStatusDetail;
+import org.openhab.core.thing.binding.BaseThingHandler;
+import org.openhab.core.thing.binding.builder.ChannelBuilder;
+import org.openhab.core.thing.binding.builder.ThingBuilder;
+import org.openhab.core.thing.type.ChannelTypeUID;
+import org.openhab.core.types.Command;
+import org.openhab.core.types.RefreshType;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.gson.Gson;
+
+/**
+ * The {@link EvccHandler} is responsible for handling commands, which are
+ * sent to one of the channels.
+ *
+ * @author Florian Hotze - Initial contribution
+ */
+@NonNullByDefault
+public class EvccHandler extends BaseThingHandler {
+    private final Logger logger = LoggerFactory.getLogger(EvccHandler.class);
+    private final Gson gson = new Gson();
+    private @Nullable ScheduledFuture<?> statePollingJob;
+
+    private @Nullable EvccConfiguration config;
+
+    private @Nullable Status status;
+
+    private int numberOfLoadpoints = 0;
+    private @Nullable String sitename;
+    private boolean batteryConfigured = false;
+    private boolean gridConfigured = false;
+    private boolean pvConfigured = false;
+
+    private boolean targetTimeEnabled = false;
+    private int targetSoC = 0;
+    private ZonedDateTime targetTimeZDT = ZonedDateTime.now();
+
+    public EvccHandler(Thing thing) {
+        super(thing);
+    }
+
+    @Override
+    public void handleCommand(ChannelUID channelUID, Command command) {
+        if (command.equals(RefreshType.REFRESH)) {
+            refresh();
+        } else {
+            logger.debug("Handling command {} for channel {}", command, channelUID);
+            String channelIdWithoutGroup = channelUID.getIdWithoutGroup();
+            int loadpoint = Integer.parseInt(channelUID.getGroupId().toString().substring(9));
+            targetSoC = status.getResult().getLoadpoints()[loadpoint].getTargetSoC();
+            switch (channelIdWithoutGroup) {
+                case CHANNEL_LOADPOINT_MODE:
+                    setMode(config.url, loadpoint, command.toString());
```

Comment:
Here and for all other set methods: Error handling is missing. The set methods returns a `@Nullable String` received from `httpRequest()`. In case of failed calls, this string appears to have the content "{\"response_code\":\"999\"}". Perhaps it would be simpler to throw an exception in this case, which could then be handled?

Reply:
I will return `null` if request failed and therefore set methods would also return `null` which then will be handled.